### ISSUE TITLE
PRD: Task-Driven Agent Evolution Gym

### ISSUE THREAD

**Issue by CCDawn** [CCDawn/Vibelution#1]
## Problem Statement

Vibelution currently has early **Self-Evolution** and **Supervised Evolution** machinery, including **Datasets**, **Bundles**, **Cases**, **Decision Records**, **Lineage**, and a **Selection Policy**. This proves that candidate behavior can be compared against baseline behavior, but the system is still mainly task-triggered and human-directed.

The next step is a task-driven evolution gym: Vibelution should improve by repeatedly doing known and generated **Cases**, scoring its own Attempts, diagnosing which part of the Agent harness limited performance, proposing one focused improvement, and accepting that improvement only when evaluation shows measurable gain without unacceptable regression.

This is not a tool-generation-only feature. New **Tools** are one possible outcome, but the broader goal is to improve the entire Agent harness: planning, context management, Tool routing, validation strategy, memory policy, recovery behavior, Workbench flows, and promotion policy.

## Solution

Build a task-driven Agent Evolution Gym that turns Vibelution's growth into a measurable closed loop:

1. Select or generate **Cases** from a **Dataset**.
2. Run the current Agent harness against those Cases.
3. Persist full Attempt traces, Tool Calls, errors, validation results, and scores.
4. Diagnose the dominant harness gap behind failures or inefficiency.
5. Propose exactly one bounded improvement to a harness module.
6. Evaluate the candidate Agent against development and holdout Bundles.
7. Accept, hold, or reject the improvement through an explicit **Selection Policy**.
8. Persist accepted learning into **Lineage**, memory, skills, policy, or code as appropriate.

The product outcome is an Agent-centered training loop. Vibelution is not waiting for a user to request improvements; it is exercising against a gym, measuring itself, and retaining only improvements that survive selection pressure.

## User Stories

1. As Vibelution, I want a named Gym that contains task-driven Cases, so that my growth is shaped by repeatable external pressure.
2. As Vibelution, I want each Case to declare objective, setup, allowed tools, scoring rules, and validation commands, so that I can attempt it without hidden human judgment.
3. As Vibelution, I want to materialize Gym Cases into Bundles, so that existing Supervised Evolution workflows can run them.
4. As Vibelution, I want to run a baseline Agent against a Bundle, so that I know my current capability before proposing changes.
5. As Vibelution, I want to record each Attempt as structured trace data, so that later diagnosis can inspect what actually happened.
6. As Vibelution, I want Tool Calls, command failures, retries, file reads, edits, tests, and final outcomes to be part of the Attempt trace, so that failures can be attributed to harness behavior rather than vague model weakness.
7. As Vibelution, I want Attempt scoring to include success, steps, elapsed time, Tool errors, validation coverage, regression count, and artifact quality, so that improvement is multi-dimensional.
8. As Vibelution, I want a Critic to classify failures by harness gap, so that I can distinguish planning problems from context, Tool routing, validation, memory, and recovery problems.
9. As Vibelution, I want the Critic to produce evidence-backed diagnoses, so that proposed improvements are grounded in trace facts.
10. As Vibelution, I want the Critic to identify repeated inefficient behavior, so that evolution can improve speed and reliability even when Cases eventually pass.
11. As Vibelution, I want an Evolver to propose one focused improvement at a time, so that A/B evaluation can attribute changes to a specific cause.
12. As Vibelution, I want improvement types to include policy patch, context patch, memory patch, Tool patch, verifier patch, scheduler patch, Workbench flow patch, and prompt patch, so that evolution is not reduced to creating new Tools.
13. As Vibelution, I want every proposed improvement to include an expected effect, so that evaluation can falsify it.
14. As Vibelution, I want candidate improvements to run in an isolated evolution transaction, so that unsafe changes cannot silently become part of the baseline.
15. As Vibelution, I want the Evaluator to compare baseline and candidate on development Bundles, so that local gains are measured before broader promotion.
16. As Vibelution, I want the Evaluator to compare baseline and candidate on holdout Bundles, so that overfitting to the current Case is detected.
17. As Vibelution, I want the Evaluator to run regression Bundles for core behavior, so that growth does not damage basic Agent safety or runtime stability.
18. As Vibelution, I want the Selection Policy to decide promote, hold, observe, or reject, so that evolution decisions are consistent and auditable.
19. As Vibelution, I want accepted improvements to update Lineage, so that future generations know which baseline they descend from.
20. As Vibelution, I want rejected improvements to be stored with their failure reason, so that I avoid rediscovering the same dead end.
21. As Vibelution, I want held improvements to enter an observation pool, so that promising but uncertain changes can be revisited.
22. As Vibelution, I want the Gym to support fixed benchmark Cases, so that progress can be compared across Agent versions.
23. As Vibelution, I want the Gym to support generated Cases, so that I can create new training pressure from observed weaknesses.
24. As Vibelution, I want generated Cases to carry source lineage, so that synthetic exercise growth remains auditable.
25. As Vibelution, I want Case tags for planning, context management, Tool routing, validation, memory, recovery, Workbench, and codebase navigation, so that performance can be tracked by capability.
26. As Vibelution, I want score dashboards by capability tag, so that the next improvement target is chosen from evidence.
27. As Vibelution, I want the Workbench to expose Gym runs, recent Attempts, Decision Records, and Lineage summaries, so that operators can inspect the evolution loop without memorizing flags.
28. As Vibelution, I want the Workbench to recover from blocked Datasets and invalid setup, so that Gym operation does not collapse on incomplete sources.
29. As Vibelution, I want the Gym to ingest public task environments such as AgentGym-style trajectories, so that growth can start from existing research assets.
30. As Vibelution, I want external trajectories to be normalized into local Cases and Attempts, so that imported data speaks Vibelution's domain language.
31. As Vibelution, I want imported traces to be marked as reference behavior rather than automatically accepted Agent behavior, so that external data does not bypass local evaluation.
32. As Vibelution, I want Reflexion-style reflection records from failed Attempts, so that language-level lessons can improve future behavior without immediate code changes.
33. As Vibelution, I want Voyager-style reusable skills to be extracted only after repeated success, so that the skill library stays useful rather than noisy.
34. As Vibelution, I want ACE-style context curation, so that Core Prompt, Dynamic Prompt, and State Memory improve without collapsing into vague summaries.
35. As Vibelution, I want Tool-Genesis-style checks when a new Tool is proposed, so that Tool schema, unit behavior, and downstream task value are evaluated separately.
36. As Vibelution, I want AgentEvolver-style self-questioning to create variants of failed Cases, so that one weakness becomes a family of exercises.
37. As Vibelution, I want every accepted improvement to include a compact post-mortem, so that future Critic decisions learn from selection outcomes.
38. As Vibelution, I want every evolution episode to be replayable, so that regressions can be reproduced.
39. As Vibelution, I want cost and runtime to be part of scoring, so that evolution favors effective and efficient harness behavior.
40. As Vibelution, I want safety constraints to be part of scoring, so that risky shortcuts cannot be mistaken for capability gains.
41. As Vibelution, I want an improvement to be rejected if it improves one tag but causes unacceptable regression on core tags, so that growth remains balanced.
42. As Vibelution, I want the Gym to support small smoke Bundles, so that harness changes can be checked quickly.
43. As Vibelution, I want the Gym to support larger scheduled evaluation Bundles, so that slower but deeper progress checks can run outside interactive work.
44. As Vibelution, I want evolution artifacts to live in the Workspace, so that runtime data remains separate from stable Core behavior.
45. As Vibelution, I want reusable evolution logic to live in Core modules, so that Workbench and CLI surfaces remain thin.
46. As Vibelution, I want issue-ready PRDs and plans to be produced from accepted research, so that large evolution work can be split into agent-executable implementation slices.
47. As Vibelution, I want the Gym to avoid direct dependence on a single public benchmark, so that the evolution loop remains useful even if one source is unavailable.
48. As Vibelution, I want public research sources to be tracked as provenance, so that future maintainers know why each mechanism exists.
49. As Vibelution, I want the first version to work with local Bundles before external environments, so that the loop becomes reliable before integrations expand.
50. As Vibelution, I want every module boundary to be testable in isolation, so that harness evolution does not create a fragile monolith.

## Implementation Decisions

- Build the feature around a task-driven evolution loop rather than a tool-generation loop. Tool creation is a supported improvement type, not the main abstraction.
- Treat the primary data unit as an improvement episode: exercise, baseline Attempt, diagnosis, candidate improvement, candidate Attempt, evaluation, and Selection Policy decision.
- Preserve the existing vocabulary of Dataset, Bundle, Case, Decision Record, Lineage, Selection Policy, Workbench, State Memory, and Workspace.
- Add a Gym module that owns Case selection, generated Case registration, Case tags, and Bundle creation for evolution exercises.
- Add an Attempt Runner module that executes a Bundle against a specific Agent version or harness variant and emits structured Attempt records.
- Add a Trace Store module that persists Tool Calls, observations, errors, validation outputs, score components, and artifacts in a queryable form.
- Add a Critic module that reads Attempt traces and classifies harness gaps using a fixed taxonomy: planning, context management, Tool routing, validation, memory, recovery, Workbench flow, execution environment, and safety policy.
- Add an Evolver module that proposes one bounded improvement from a Critic diagnosis. It should emit an improvement spec before any code or prompt mutation is applied.
- Add an Evaluator module that compares baseline and candidate behavior on development, holdout, smoke, and regression Bundles.
- Extend the Selection Policy so that it can evaluate full harness improvements, not only candidate behavior on a single supervised run.
- Keep Workbench as a thin shell. It should ask deeper evolution modules for available Gyms, Bundles, Attempts, Decision Records, and Lineage summaries.
- Treat imported public research data as seed material. AgentGym-like traces can become Cases and reference traces, but local acceptance must still require Vibelution evaluation.
- Use AgentGym as the design reference for multi-environment exercise structure and trajectory capture.
- Use Voyager as the design reference for reusable skill extraction and long-term capability accumulation.
- Use Reflexion as the design reference for failed Attempt reflection records that can improve future behavior without immediate code mutation.
- Use AgentEvolver as the design reference for self-generated exercises and credit assignment across long traces.
- Use Tool-Genesis as the design reference for evaluating generated Tools across interface correctness, functional correctness, and downstream usefulness.
- Use ACE as the design reference for context and playbook evolution with curation, anti-collapse checks, and provenance.
- The core improvement episode schema should be stable enough to support future training. A decision-rich shape is:

```json
{
  "exercise": {
    "case_id": "string",
    "bundle_name": "string",
    "capability_tags": ["planning", "validation"],
    "success_metric": "string"
  },
  "baseline_attempt": {
    "agent_version": "string",
    "score": {},
    "trace_id": "string"
  },
  "diagnosis": {
    "harness_gap": "planning | context | tool_routing | validation | memory | recovery | workbench | environment | safety",
    "evidence": ["string"]
  },
  "improvement": {
    "type": "policy_patch | context_patch | memory_patch | tool_patch | verifier_patch | scheduler_patch | workbench_patch | prompt_patch",
    "expected_effect": "string"
  },
  "candidate_attempt": {
    "agent_version": "string",
    "score": {},
    "trace_id": "string"
  },
  "selection": {
    "decision": "PROMOTE | HOLD | OBSERVE | REJECT",
    "reason": "string"
  }
}
```

- The first release should prefer local, deterministic Cases and Bundles over heavy external environments. External integrations can follow once the loop is reliable.
- Generated Cases must never automatically count as holdout evaluation. Holdout Bundles should be stable and versioned.
- Accepted improvements should update Lineage and write an auditable Decision Record. Rejected and held improvements should also be retained as learning artifacts.

## Testing Decisions

- Tests should verify observable evolution behavior rather than private implementation details.
- Gym tests should verify that Cases can be selected, tagged, and materialized into Bundles with stable metadata.
- Attempt Runner tests should verify that a Bundle run produces Attempt records with scores, trace identifiers, and validation artifacts.
- Trace Store tests should verify round-trip persistence and filtering by Case, Bundle, Agent version, capability tag, and decision.
- Critic tests should use small synthetic traces to verify that common failure patterns map to the expected harness gap taxonomy.
- Evolver tests should verify that a diagnosis produces exactly one bounded improvement spec with an expected effect and target module type.
- Evaluator tests should verify baseline-versus-candidate scoring, holdout regression detection, and cost-aware comparison.
- Selection Policy tests should verify promote, hold, observe, and reject decisions for multi-metric improvement episodes.
- Workbench tests should verify external behavior: listing Gyms, selecting Bundles, running evaluation, recovering from blocked setup, showing recent Decision Records, and showing Lineage summaries.
- Regression tests should reuse prior supervised-evolution tests as precedent: persist Decision Records, update Lineage, reject regressions, hold uncertain improvements, and promote only when gates pass.
- Public data ingestion tests should use tiny fixtures, not live downloads, so that test runs remain deterministic.
- End-to-end smoke tests should run a minimal local Gym Bundle and assert that a complete improvement episode can be recorded without requiring a real harness mutation.

## Out of Scope

- Training or fine-tuning a model's weights is out of scope for the first release.
- Full SWE-bench harness integration is out of scope for the first release.
- Browser, Minecraft, WebShop, or other heavy external environments are out of scope for the first release.
- Autonomous unbounded self-modification is out of scope. All candidate improvements must pass explicit evaluation and Selection Policy gates.
- Replacing existing Supervised Evolution is out of scope. The Gym should build on it and generalize it.
- Building a public benchmark service is out of scope.
- Automatically trusting public trajectories as accepted Vibelution behavior is out of scope.
- Optimizing for a single score while ignoring safety, cost, and regression is out of scope.

## Further Notes

- The research synthesis suggests using AgentGym for exercise and trajectory concepts, Voyager for reusable skill accumulation, Reflexion for language-level learning from failed Attempts, AgentEvolver for self-generated training pressure, Tool-Genesis for Tool creation evaluation, and ACE for context/playbook evolution.
- The first milestone should prove a local closed loop with a small fixed Gym Bundle and no external services.
- The second milestone should add Critic-driven improvement specs and multi-metric evaluation.
- The third milestone should introduce generated Cases and observation pools.
- The fourth milestone should import selected public trajectory data into local Case format.
- The long-term goal is measurable task-driven Self-Evolution: Vibelution repeatedly improves its ability to solve task distributions while preserving safety, stability, and core behavior.
